pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+//! It verifies headers submitted to it via on-chain transactions, performed by a so-called relayer
+//! It is a part of the Aleph Zero <-> Terra bridge
+pub use pallet::*;
+
+pub mod types;
+mod utils;
+
+use frame_support::traits::StorageVersion;
+
+/// The current storage version.
+const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use crate::types::{
+        LightBlockStorage, LightClientOptionsStorage, TendermintBlockHash, TendermintHashStorage,
+    };
+    use frame_support::{
+        ensure, fail, log,
+        pallet_prelude::{
+            DispatchClass, DispatchResult, IsType, OptionQuery, StorageMap, StorageValue,
+            ValueQuery,
+        },
+        traits::{Get, UnixTime},
+        Identity,
+    };
+    use frame_system::{ensure_root, ensure_signed, pallet_prelude::OriginFor};
+    use tendermint::Time;
+    use tendermint_light_client_verifier::{
+        options::Options, types::LightBlock, ProdVerifier, Verdict, Verifier,
+    };
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// ubiquitous event type
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Maximal number of finalized headers to keep in the storage, last-in first-out
+        #[pallet::constant]
+        type HeadersToKeep: Get<u32>;
+
+        /// Maximal number of block validators in Tendermint
+        #[pallet::constant]
+        type MaxVotesCount: Get<u32>;
+
+        /// time provider type, used to gauge whether blocks are within the trusting period
+        type TimeProvider: UnixTime;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Pallet is halted
+        LightClientHalted,
+        /// Pallet operations are resumed        
+        LightClientResumed,
+        /// Light client is initialized
+        LightClientInitialized,
+        /// New block has been verified and imported into storage \[relayer_address, imported_block_hash\]
+        ImportedLightBlock(T::AccountId, TendermintBlockHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Light client has not been initialized        
+        NotInitialized,
+        /// Light client has already been initialized
+        AlreadyInitialized,
+        /// Light client is currently halted
+        Halted,
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid        
+        InvalidBlock,
+        /// Initial block is invalid
+        InvalidInitialBlock,
+        /// General error during client operations
+        Other,
```

Comment:
Can we enrich this error variant with a message? If not, do we log the error? Trying to think how the consumer of it can figure out what went wrong.

Reply:
It would be great to add message String here, but the pallet::error must be of Unit type :/ in Substarte. 

We do log runtime errors however, and the ConversionError from types mod preserves all context, so we will be able to sift through logs and see what went wrong and where.

Reply:
We will, I was wondering what would other users of the chain do. They won't have access to the chain.

I understand that we can't add message "because Substrate".

Reply:
I return Error<T>::ConversionError in addition to Other now